### DISCUSSION TITLE
Need Help to Export data to the Secured HTPP Endpoint

### DISCUSSION THREAD

**Discussion by kumardeepak5** [edgexfoundry/community#191] [Q&A]
Hi Team,

I have requirement to export device data to the secured https endpoint.

**Below is description of my REST Endpoint-**
I need to pass an Bearer token (for authentication & authorization) to call my REST Endpoint where i have to write device data.
To Get Bearer token i need to call another rest endpoint which provides Bearer token after parsing the response.

**My Queries**
- Please let me know can i implement same by using App-service-configurable.
- Or i need to implement my custom app service by using app-function-sdk-go.
- If i need to create custom service then can you please guide me for same where actually i need to make code change?

Thanks
Deepak Kumar

**Comment by lenny-goodell** [comment]
@kumardeepak5 , this will not be possible with App-service-configurable due to the need to "call another rest endpoint which provides Bearer token after parsing the response". You will need to create a custom app service.

Please read the App Service documentation here to familiarize yourself with creating custom application services.
https://docs.edgexfoundry.org/3.0/microservices/application/ApplicationServices/
There are many examples here: https://github.com/edgexfoundry/edgex-examples/tree/v3.0/application-services/custom
You can use the quick start [App Service template](https://github.com/edgexfoundry/app-functions-sdk-go/blob/v3.0.1/app-service-template/README.md) to create your new custom App Service. 

Your custom app service will need a custom pipeline function that "calls another rest endpoint which provides Bearer token after parsing the response". At this point your custom pipeline function can store the `Bearer token` in the `SecretStore` via the SecretProvider on the App Function Context. See https://docs.edgexfoundry.org/3.0/microservices/application/AppFunctionContextAPI/#secretprovider.

If you are running EdgeX in secure mode, this will be stored securely in Vault, otherwise this will be stored insecurely in service's InsecureSecrets configuration section in Config Provider (i.e. Consul).

Then all you need to do is configure the built-in HTTP Export to use the use your `Bearer token` from the `SecretStore` using NewHTTPSenderWithSecretHeader() factory function. See here for details: https://docs.edgexfoundry.org/3.0/microservices/application/BuiltIn/#http-export

If you are running non-secure mode and don't like the `Bearer token` stored insecurely in Consul, you can store it in the App Function Context Storage (memory) for a custom export function to extract and use. See here for more details on App Function Context Storage: https://docs.edgexfoundry.org/3.0/microservices/application/AppFunctionContextAPI/#context-storage
In this case you have to aslo write a simple custom HTTP Export pipeline function that knows to get the  `Bearer token` from App Function Context Storage and uses it in the HTTPS header.

Does the `endpoint which provides Bearer token` need to be called every time? or only the first time it is needed? or also when the Bearer token has expired? The recommendation above assumes it is called every time, but you could check if it already exist in SecretStore/Context Storage before calling the endpoint and skipping the call if you have already loaded it into SecretStore/Context Storage.
